File: test_sort.js
function sortTestRows(rows) {
  const parseSequence = (name) => {
    const match = name.match(/^([A-Za-z\-]+)(\d+)(.*)$/);
    if (!match) return null;
    return {
      prefix: match[1].toUpperCase().replace(/[^A-Z]/g, ''),
      num: parseInt(match[2], 10)
    };
  };

  const getLogicalIndex = (seq) => {
    if (!seq) return -1;
    if (seq.prefix === 'MT' || seq.prefix === 'PT') return seq.num * 10;
    if (seq.prefix === 'CMT' || seq.prefix === 'JCT') return seq.num * 20 + 5;
    if (seq.prefix === 'FMT') return 1000 + seq.num * 10;
    return -1;
  };

  return rows.sort((a, b) => {
    const seqA = parseSequence(a.name);
    const seqB = parseSequence(b.name);
    const idxA = getLogicalIndex(seqA);
    const idxB = getLogicalIndex(seqB);

    if (idxA !== -1 && idxB !== -1 && idxA !== idxB) {
      return idxA - idxB;
    }
    return a.name.localeCompare(b.name, undefined, { numeric: true });
  });
}

const tests = [
  { name: 'MT02' },
  { name: 'CMT01' },
  { name: 'MT01' },
  { name: 'MT03' },
  { name: 'CMT02' },
  { name: 'FMT01' },
  { name: 'MT04' }
];

console.log(sortTestRows(tests).map(t => t.name));

File: test_sort.py
import re

def parse_sequence(name):
    match = re.match(r"^([A-Za-z\-]+)(\d+)(.*)$", name)
    if not match:
        return None
    prefix = re.sub(r'[^A-Z]', '', match.group(1).upper())
    num = int(match.group(2))
    return {"prefix": prefix, "num": num}

def get_logical_index(seq):
    if not seq:
        return -1
    prefix = seq["prefix"]
    num = seq["num"]
    if prefix in ['MT', 'PT']:
        return num * 10
    if prefix in ['CMT', 'JCT']:
        return num * 20 + 5
    if prefix == 'FMT':
        return 1000 + num * 10
    return -1

def sort_key(test):
    name = test["name"]
    seq = parse_sequence(name)
    idx = get_logical_index(seq)
    
    # We can return a tuple for sorting: (idx if found else float('inf'), name)
    if idx != -1:
        return (0, idx, name)
    return (1, 0, name)

tests = [
  { "name": 'MT02' },
  { "name": 'CMT01' },
  { "name": 'MT01' },
  { "name": 'MT03' },
  { "name": 'CMT02' },
  { "name": 'FMT01' },
  { "name": 'MT04' }
]

tests.sort(key=sort_key)
print([t["name"] for t in tests])
